syntax = "proto3";

enum PbDeflateFlush {
  PB_DEFLATE_Z_NO_FLUSH = 0;
  PB_DEFLATE_Z_PARTIAL_FLUSH = 1;
  PB_DEFLATE_Z_SYNC_FLUSH = 2;
  PB_DEFLATE_Z_FULL_FLUSH = 3;
  PB_DEFLATE_Z_BLOCK = 5;
}

enum PbInflateFlush {
  PB_INFLATE_Z_NO_FLUSH = 0;
}

enum PbLevel {
  PB_Z_NO_COMPRESSION = 0;
  PB_Z_BEST_SPEED = 1;
  PB_Z_LEVEL2 = 2;
  PB_Z_LEVEL3 = 3;
  PB_Z_LEVEL4 = 4;
  PB_Z_LEVEL5 = 5;
  PB_Z_LEVEL6 = 6;
  PB_Z_LEVEL7 = 7;
  PB_Z_LEVEL8 = 8;
  PB_Z_BEST_COMPRESSION = 9;
  PB_Z_DEFAULT_COMPRESSION = -1;
}

enum PbStrategy {
  PB_Z_DEFAULT_STRATEGY = 0;
  PB_Z_FILTERED = 1;
  PB_Z_HUFFMAN_ONLY = 2;
  PB_Z_RLE = 3;
  PB_Z_FIXED = 4;
}

enum PbWindowBits {
  PB_WB_DEFAULT = 0;
  PB_WB_RAW = -15;
  PB_WB_ZLIB = 15;
  PB_WB_GZIP = 31;
}

enum PbMemLevel {
  PB_MEM_LEVEL_DEFAULT = 0;
  PB_MEM_LEVEL1 = 1;
  PB_MEM_LEVEL2 = 2;
  PB_MEM_LEVEL3 = 3;
  PB_MEM_LEVEL4 = 4;
  PB_MEM_LEVEL5 = 5;
  PB_MEM_LEVEL6 = 6;
  PB_MEM_LEVEL7 = 7;
  PB_MEM_LEVEL8 = 8;
  PB_MAX_MEM_LEVEL = 9;
}

message PbDeflate {
  int32 avail_in = 1;
  int32 avail_out = 2;
  PbDeflateFlush flush = 3;
}

message PbDeflateParams {
  int32 avail_in = 1;
  int32 avail_out = 2;
  PbLevel level = 3;
  PbStrategy strategy = 4;
}

message PbInflate {
  int32 avail_in = 1;
  int32 avail_out = 2;
  PbInflateFlush flush = 3;
}

message PbDeflateOp {
  oneof op {
    PbDeflate deflate = 1;
    PbDeflateParams deflate_params = 2;
  }
}

message PbInflateOp {
  oneof op {
    PbInflate inflate = 1;
  }
}

message PbPlan {
  bytes data = 1;
  PbLevel level = 2;
  PbWindowBits window_bits = 3;
  PbMemLevel mem_level = 4;
  PbStrategy strategy = 5;
  bytes dict = 6;
  repeated PbDeflateOp deflate_ops = 7;
  repeated int32 finish_avail_outs = 8;
  repeated PbInflateOp inflate_ops = 9;
  int32 tail_size = 10;
  repeated int32 bit_flips = 11;
}
